test/unit/org/apache/cassandra/net/ConnectionUtils.java

```diff
@@ -197,45 +197,42 @@ public InboundCountChecker error(long count, long bytes)
+                testAndFailCheck.accept("received count values don't match", received, connection::receivedCount);
+                testAndFailCheck.accept("received bytes values don't match", receivedBytes, connection::receivedBytes);
             }
             if (checkProcessed)
             {
-                testAndFailCheck.accept("processed count values don't match", processed, connection.processedCount());
-                testAndFailCheck.accept("processed bytes values don't match", processedBytes, connection.processedBytes());
+                testAndFailCheck.accept("processed count values don't match", processed, connection::processedCount);
+                testAndFailCheck.accept("processed bytes values don't match", processedBytes, connection::processedBytes);
             }
             if (checkExpired)
             {
-                testAndFailCheck.accept("expired count values don't match", expired, connection.expiredCount());
-                testAndFailCheck.accept("expired bytes values don't match", expiredBytes, connection.expiredBytes());
+                testAndFailCheck.accept("expired count values don't match", expired, connection::expiredCount);
+                testAndFailCheck.accept("expired bytes values don't match", expiredBytes, connection::expiredBytes);
             }
             if (checkError)
             {
-                testAndFailCheck.accept("error count values don't match", error, connection.errorCount());
-                testAndFailCheck.accept("error bytes values don't match", errorBytes, connection.errorBytes());
+                testAndFailCheck.accept("error count values don't match", error, connection::errorCount);
+                testAndFailCheck.accept("error bytes values don't match", errorBytes, connection::errorBytes);
             }
             if (checkScheduled)
             {
                 // scheduled cannot relied upon to not race with completion of the task,
-                // so if it is currently above the value we expect, sleep for a bit
-                if (scheduled < connection.scheduledCount())
-                    for (int i = 0; i < 10 && scheduled < connection.scheduledCount() ; ++i)
-                        Uninterruptibles.sleepUninterruptibly(1L, TimeUnit.MILLISECONDS);
-                testAndFailCheck.accept("scheduled count values don't match", scheduled, connection.scheduledCount());
-                testAndFailCheck.accept("scheduled bytes values don't match", scheduledBytes, connection.scheduledBytes());
+                // so if it is currently above the value we expect, use the spinAssert explicitly.
+                spinAssertEquals("scheduled count values don't match", scheduled, connection::scheduledCount, 5, TimeUnit.SECONDS);
+                spinAssertEquals("scheduled bytes values don't match", scheduledBytes, connection::scheduledBytes, 5, TimeUnit.SECONDS);
```

Comment:
NIT: I would have used `testAndFailCheck` for ensuring that the behavior is the one expected by the `FailCheck` argument.
In practice, unless I missed something, this code is only called from `ConnectionTest` so there will be no change of behavior.

I can change that on commit if you want to.

Reply:
It is probably better to use `spinAssertEquals` explicitly. If you strongly prefer to use `testAndFailCheck`, I am cool with it.

The spin equality assertion was used because 
- the original code that Benedict wrote did it explicitly using a for loop. The assertion is likely to require multiple retries. 
- the parameter type `FailCheck` for `doCheck()` suggests it can accept any `FailCheck` impl. We cannot assume the input check is always spin assertion based (, although it is only used in `ConnectionTest` as of now). If a custom non-spin assertion is passed, the check can fail.
- the error message, i.e. "scheduled count values don't match", suggests the purpose is to assert values are equal. So `spinAssertEquals` does the job.

Reply:
If all those failing tests have proven us somthing is that the spin would be needed everywhere. The `doCheck` method with the FailCheck parameter is used in the burn tests to allow the implementor to plug his own notification mechanism for failure. That would not work with the current implementation that use the spin assert.
My approach would be to let the user of that method handle the spinning if needed but I agree it is not a perfect solution.
Otherwise we need to implements a spining that will work for `doCheck()`and `doCheck(FailCheck)`.